yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,119 @@ namespace flowable {
+  void onError(folly::exception_wrapper) override {
+    DCHECK(subscription_) << "Calling onError() without a subscription";
+    subscription_.reset();
+  }
+
+ protected:
+  Reference<Subscription> subscription() {
+    return subscription_;
+  }
+
+ private:
+  Reference<Subscription> subscription_;
+};
+
+// T : Type of Flowable that this Subscriber operates on
+// keeps_reference_to_this : Indicates that the subclass will ensure
+// that a reference to `this` will stay live as long as `this` can
+// be accessed (eg understands that it's possible to drop all references
+// to the object within a method of that object)
+// Classes that ensure this can set keeps_reference_to_this to `true`
+// as an optimization (to prevent an atomic inc/dec pair)
+template <typename T, bool keeps_reference_to_this = false>
+class SafeSubscriber : public Subscriber<T> {
+ public:
   // Note: If any of the following methods is overridden in a subclass, the new
   // methods SHOULD ensure that these are invoked as well.
-  virtual void onSubscribe(Reference<Subscription> subscription) {
+  void onSubscribe(Reference<Subscription> subscription) final override {
     DCHECK(subscription);
+    CHECK(!subscription_);
 
-    if (subscription_) {
-      subscription->cancel();
-      return;
+    Reference<SafeSubscriber> self;
+    if(!keeps_reference_to_this) {
+      self = ref_from_this(this);
     }
 
     subscription_ = std::move(subscription);
+    onSubscribeImpl();
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onComplete() {
+  void onComplete() final override {
     DCHECK(subscription_) << "Calling onComplete() without a subscription";
-    subscription_.reset();
+
+    Reference<SafeSubscriber> self;
+    if(!keeps_reference_to_this) {
+      self = ref_from_this(this);
+    }
+    auto sub = std::move(subscription_);
+
+    onCompleteImpl();
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onError(folly::exception_wrapper) {
+  void onError(folly::exception_wrapper e) final override {
     DCHECK(subscription_) << "Calling onError() without a subscription";
-    subscription_.reset();
+
+    Reference<SafeSubscriber> self;
+    if(!keeps_reference_to_this) {
+      self = ref_from_this(this);
+    }
+    auto sub = std::move(subscription_);
+
+    onErrorImpl(std::move(e));
   }
 
-  virtual void onNext(T) = 0;
+  void onNext(T t) final override {
+    Reference<SafeSubscriber> self;
+    if(!keeps_reference_to_this) {
+      self = ref_from_this(this);
+    }
+
+    onNextImpl(std::move(t));
+  }
+
+  virtual void onSubscribeImpl() = 0;
+  virtual void onCompleteImpl() = 0;
+  virtual void onNextImpl(T) = 0;
+  virtual void onErrorImpl(folly::exception_wrapper) = 0;
 
  protected:
-  Subscription* subscription() {
-    return subscription_.operator->();
+  Reference<Subscription> subscription() {
```

Comment:
Ownership of the `subscription_` seems like an issue that makes the job of classes that extends Subscriber harder. Can we have a function to detach the `subscription_` from the parent class if child class wants to own the object.

Like, we call `onSubscribe(void)`, but the child class want to force to have the ownership of the `subscription_`, so instead of calling `subscription()` function, it wants `detachSubscription() : Reference<Subscription>`, which gets rid of any reference in the parent class.